gazebo_ros/src/gazebo_ros_force_system.cpp

```diff
@@ -0,0 +1,422 @@
+  rclcpp::Service<gazebo_msgs::srv::JointRequest>::SharedPtr clear_joint_efforts_service_;
+
+  /// To be notified once the world is created.
+  gazebo::event::ConnectionPtr world_created_connection_;
+
+  /// Connection to event called at every world iteration.
+  gazebo::event::ConnectionPtr update_connection_;
+
+  /// Pointer to apply wrench tasks
+  std::vector<std::shared_ptr<LinkWrenchTask>> link_wrench_tasks_;
+
+  /// Pointer to joint effort tasks
+  std::vector<std::shared_ptr<JointEffortTask>> joint_effort_tasks_;
+
+  /// Lock for executing tasks
+  std::mutex lock_;
+};
+
+GazeboRosForceSystem::GazeboRosForceSystem()
+: impl_(std::make_unique<GazeboRosForceSystemPrivate>())
+{
+}
+
+GazeboRosForceSystem::~GazeboRosForceSystem()
+{
+}
+
+void GazeboRosForceSystem::Load(int /* argc */, char ** /* argv */)
+{
+  impl_->update_connection_ = gazebo::event::Events::ConnectWorldUpdateBegin(
+    std::bind(&GazeboRosForceSystemPrivate::TaskExecutor, impl_.get(), std::placeholders::_1));
+
+  impl_->world_created_connection_ = gazebo::event::Events::ConnectWorldCreated(
+    std::bind(&GazeboRosForceSystemPrivate::OnWorldCreated, impl_.get(), std::placeholders::_1));
+}
+
+void GazeboRosForceSystemPrivate::OnWorldCreated(const std::string & _world_name)
+{
+  // Only support one world
+  world_created_connection_.reset();
+
+  world_ = gazebo::physics::get_world(_world_name);
+
+  // ROS transport
+  ros_node_ = gazebo_ros::Node::Get();
+
+  apply_link_wrench_service_ = ros_node_->create_service<gazebo_msgs::srv::ApplyLinkWrench>(
+    "apply_link_wrench", std::bind(&GazeboRosForceSystemPrivate::ApplyLinkWrench, this,
+    std::placeholders::_1, std::placeholders::_2));
+
+  clear_link_wrenches_service_ = ros_node_->create_service<gazebo_msgs::srv::LinkRequest>(
+    "clear_link_wrenches", std::bind(&GazeboRosForceSystemPrivate::ClearLinkWrenches, this,
+    std::placeholders::_1, std::placeholders::_2));
+
+  apply_joint_effort_service_ = ros_node_->create_service<gazebo_msgs::srv::ApplyJointEffort>(
+    "apply_joint_effort", std::bind(&GazeboRosForceSystemPrivate::ApplyJointEffort, this,
+    std::placeholders::_1, std::placeholders::_2));
+
+  clear_joint_efforts_service_ = ros_node_->create_service<gazebo_msgs::srv::JointRequest>(
+    "clear_joint_efforts", std::bind(&GazeboRosForceSystemPrivate::ClearJointEfforts, this,
+    std::placeholders::_1, std::placeholders::_2));
+}
+
+void GazeboRosForceSystemPrivate::TaskExecutor(const gazebo::common::UpdateInfo & _info)
+{
+  // Store sim time
+  rclcpp::Time sim_time = Convert<builtin_interfaces::msg::Time>(_info.simTime);
+
+  std::lock_guard<std::mutex> scoped_lock(lock_);
+
+  for (auto link_wrench_task = link_wrench_tasks_.begin();
+    link_wrench_task != link_wrench_tasks_.end(); ++link_wrench_task)
+  {
+    auto start = (*link_wrench_task)->start_time;
+    auto duration = (*link_wrench_task)->duration;
+    auto end = start + duration;
```

Comment:
`start == zero` would be true even if the actual start value is 0. This would lead to the force been applied infinitely. Is there any way to check if the user has passed any value for a particular element?

Reply:
Also is there a reason for `start = max(start, sim_time)` and not `start = sim_time`

Reply:
> Is there any way to check if the user has passed any value for a particular element?